File: backend/app/api/v1/model_providers.py
# backend/app/api/v1/model_providers.py
"""模型提供商管理 API（docs/TECHv1.1.md §5.1 / PRD v1.1 §2.1）。

端点：
- GET    /api/v1/model-providers                          提供商列表摘要
- POST   /api/v1/model-providers                          创建提供商（默认自动获取模型列表）
- GET    /api/v1/model-providers/{provider_id}            提供商详情（api_keys 脱敏）
- PATCH  /api/v1/model-providers/{provider_id}            更新提供商（名称/base_url/api_keys/models）
- DELETE /api/v1/model-providers/{provider_id}            删除提供商（解除项目/会话引用）
- POST   /api/v1/model-providers/{provider_id}/fetch-models  触发获取模型列表
- POST   /api/v1/model-providers/{provider_id}/detect     检测所有 Key 连接状态
- POST   /api/v1/model-providers/{provider_id}/keys/{key_id}/detect  检测单个 Key

错误处理：提供商 / Key 不存在 → 404；参数不合法（空名称 / 非法 type / 新增 Key 缺明文）→ 400。
"""
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from ...database import get_db
from ...models import ModelProvider
from ...schemas.model_provider import (
    KeyDetectResult,
    ModelFetchResult,
    ModelProviderCreate,
    ModelProviderCreateResponse,
    ModelProviderRead,
    ModelProviderUpdate,
    ProviderSummary,
)
from ...services import model_provider as model_provider_service
from ...services.model_provider import (
    ModelProviderNotFound,
    ModelProviderValidationError,
)

router = APIRouter(prefix="/api/v1", tags=["model_providers"])

# 创建成功后自动获取模型列表失败时的提示文案（PRD v1.1 §2.1）
AUTO_FETCH_FAIL_MESSAGE = "未获取到模型"


async def _get_provider_or_404(db: AsyncSession, provider_id: str) -> ModelProvider:
    """按 id 查询提供商，不存在则抛 404。"""
    provider = await db.get(ModelProvider, provider_id)
    if provider is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"模型提供商 {provider_id} 不存在",
        )
    return provider


@router.get(
    "/model-providers",
    response_model=list[ProviderSummary],
    summary="提供商列表摘要（id/name/type/key_count/model_count/is_default/status）",
)
async def list_model_providers(db: AsyncSession = Depends(get_db)) -> list:
    """列出所有提供商摘要，不含任何 Key 信息（明文或密文）。"""
    return await model_provider_service.list_providers(db)


@router.post(
    "/model-providers",
    response_model=ModelProviderCreateResponse,
    status_code=status.HTTP_201_CREATED,
    summary="创建提供商（默认自动获取模型列表）",
)
async def create_model_provider(
    payload: ModelProviderCreate,
    db: AsyncSession = Depends(get_db),
) -> dict:
    """创建提供商并加密各 API Key；成功后自动获取模型列表。

    获取失败不阻断创建：provider 返回成功但 models 为空，并附带 message「未获取到模型」。
    """
    try:
        provider = await model_provider_service.create_provider(
            db,
            name=payload.name,
            type=payload.type,
            base_url=payload.base_url,
            api_keys=[key.model_dump() for key in payload.api_keys],
        )
    except ModelProviderValidationError as exc:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)
        ) from exc

    response: dict = {}
    if payload.auto_fetch:
        result = await model_provider_service.fetch_model_list(db, provider.id)
        response["auto_fetch"] = result
        if not result["success"]:
            response["message"] = AUTO_FETCH_FAIL_MESSAGE
    # 详情必须在 fetch 之后取，保证 models / available_models 反映最新结果
    response["provider"] = await model_provider_service.get_provider(db, provider.id)
    return response


@router.get(
    "/model-providers/{provider_id}",
    response_model=ModelProviderRead,
    summary="提供商详情（api_keys 脱敏）",
)
async def get_model_provider(
    provider_id: str,
    db: AsyncSession = Depends(get_db),
) -> dict:
    try:
        return await model_provider_service.get_provider(db, provider_id)
    except ModelProviderNotFound as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)
        ) from exc


@router.patch(
    "/model-providers/{provider_id}",
    response_model=ModelProviderRead,
    summary="更新提供商（名称/base_url/api_keys/models）",
)
async def update_model_provider(
    provider_id: str,
    payload: ModelProviderUpdate,
    db: AsyncSession = Depends(get_db),
) -> dict:
    """仅更新显式传入的字段；api_keys / models 为一次全量替换。"""
    await _get_provider_or_404(db, provider_id)
    # model_dump 已把嵌套的 ApiKeyInput / ModelItem 转为 dict，直接透传给服务层
    updates = payload.model_dump(exclude_unset=True)
    try:
        await model_provider_service.update_provider(db, provider_id, **updates)
    except ModelProviderValidationError as exc:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)
        ) from exc
    return await model_provider_service.get_provider(db, provider_id)


@router.delete(
    "/model-providers/{provider_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="删除提供商（解除项目/会话引用）",
)
async def delete_model_provider(
    provider_id: str,
    db: AsyncSession = Depends(get_db),
) -> None:
    """删除提供商：服务层先解除项目 / 会话 / 生成记录引用并清空全局默认，再删除本身。"""
    try:
        await model_provider_service.delete_provider(db, provider_id)
    except ModelProviderNotFound as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)
        ) from exc


@router.post(
    "/model-providers/{provider_id}/fetch-models",
    response_model=ModelFetchResult,
    summary="获取模型列表（合并去重所有启用 Key）",
)
async def fetch_models(
    provider_id: str,
    db: AsyncSession = Depends(get_db),
) -> dict:
    """使用所有启用 Key 请求 /models，合并去重并持久化到提供商与各 Key。"""
    try:
        return await model_provider_service.fetch_model_list(db, provider_id)
    except ModelProviderNotFound as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)
        ) from exc


@router.post(
    "/model-providers/{provider_id}/detect",
    response_model=list[KeyDetectResult],
    summary="检测所有 Key 连接状态",
)
async def detect_provider_keys(
    provider_id: str,
    db: AsyncSession = Depends(get_db),
) -> list:
    """逐个检测所有 Key 的连接状态，并刷新成功 Key 的模型列表。"""
    try:
        return await model_provider_service.detect_provider(db, provider_id)
    except ModelProviderNotFound as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)
        ) from exc


@router.post(
    "/model-providers/{provider_id}/keys/{key_id}/detect",
    response_model=KeyDetectResult,
    summary="检测单个 Key 连接状态",
)
async def detect_single_key(
    provider_id: str,
    key_id: str,
    db: AsyncSession = Depends(get_db),
) -> dict:
    """检测单个 Key 的有效性并刷新其模型列表；Key 不存在 → 404。"""
    provider = await _get_provider_or_404(db, provider_id)
    if not any(k.get("key_id") == key_id for k in provider.api_keys_json or []):
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"API Key {key_id} 不存在",
        )
    return await model_provider_service.detect_single_key(db, provider_id, key_id)

File: frontend/src/api/modelProviders.ts
import client from './client'
import type {
  ApiKeyInfo,
  ApiKeyRead,
  KeyDetectResult,
  ModelFetchResult,
  ModelItem,
  ModelProvider,
  ModelProviderCreate,
  ModelProviderCreateResponse,
  ModelProviderType,
  ModelProviderUpdate,
  ProviderStatus,
  ProviderSummary,
} from '@/types'

/**
 * 模型提供商管理 API（docs/TECHv1.1.md §5.1，对应后端 api/v1/model_providers.py）。
 * 后端响应为 snake_case JSON，此处映射为前端 camelCase 类型（见 types/modelProvider.ts）。
 */

function toApiKeyRead(raw: Record<string, unknown>): ApiKeyRead {
  return {
    keyId: raw.key_id as string,
    keyMasked: raw.key_masked as string,
    enabled: raw.enabled as boolean,
    priority: raw.priority as number,
    availableModels: Array.isArray(raw.available_models) ? (raw.available_models as string[]) : [],
  }
}

function toModelItem(raw: Record<string, unknown>): ModelItem {
  return {
    modelId: raw.model_id as string,
    enabled: raw.enabled as boolean,
    supports1mContext: raw.supports_1m_context === true,
  }
}

/** 后端 type 字符串 → 前端联合类型；未知值回退 other 保证类型安全 */
function toProviderType(raw: unknown): ModelProviderType {
  const known: ModelProviderType[] = [
    'openai',
    'anthropic',
    'deepseek',
    'kimi',
    'opencode_go',
    'custom',
    'other',
  ]
  return known.includes(raw as ModelProviderType) ? (raw as ModelProviderType) : 'other'
}

function toStatus(raw: unknown): ProviderStatus {
  return raw === 'ready' ? 'ready' : 'no_keys'
}

function toModelProvider(raw: Record<string, unknown>): ModelProvider {
  return {
    id: raw.id as string,
    name: raw.name as string,
    type: toProviderType(raw.type),
    baseUrl: (raw.base_url as string | null) ?? null,
    scope: (raw.scope as string) ?? 'global',
    isDefault: raw.is_default as boolean,
    apiKeys: Array.isArray(raw.api_keys)
      ? raw.api_keys.map((k) => toApiKeyRead(k as Record<string, unknown>))
      : [],
    models: Array.isArray(raw.models)
      ? raw.models.map((m) => toModelItem(m as Record<string, unknown>))
      : [],
    ...(raw.key_count != null ? { keyCount: raw.key_count as number } : {}),
    ...(raw.model_count != null ? { modelCount: raw.model_count as number } : {}),
    ...(raw.status != null ? { status: toStatus(raw.status) } : {}),
    createdAt: raw.created_at as string,
    updatedAt: raw.updated_at as string,
  }
}

function toProviderSummary(raw: Record<string, unknown>): ProviderSummary {
  return {
    id: raw.id as string,
    name: raw.name as string,
    type: toProviderType(raw.type),
    scope: (raw.scope as string) ?? 'global',
    baseUrl: (raw.base_url as string | null) ?? null,
    isDefault: raw.is_default as boolean,
    keyCount: (raw.key_count as number) ?? 0,
    enabledKeyCount: (raw.enabled_key_count as number) ?? 0,
    modelCount: (raw.model_count as number) ?? 0,
    enabledModelCount: (raw.enabled_model_count as number) ?? 0,
    status: toStatus(raw.status),
    createdAt: raw.created_at as string,
    updatedAt: raw.updated_at as string,
  }
}

function toFetchResult(raw: Record<string, unknown>): ModelFetchResult {
  return {
    success: raw.success as boolean,
    models: Array.isArray(raw.models) ? (raw.models as string[]) : [],
    errors: Array.isArray(raw.errors)
      ? (raw.errors as Array<{ key_id?: string; keyId?: string; error: string }>).map((e) => ({
          keyId: (e.key_id ?? e.keyId) as string,
          error: e.error,
        }))
      : [],
  }
}

function toDetectResult(raw: Record<string, unknown>): KeyDetectResult {
  return {
    keyId: raw.key_id as string,
    valid: raw.valid as boolean,
    error: (raw.error as string | null) ?? null,
    modelCount: (raw.model_count as number) ?? 0,
  }
}

/** ApiKeyInfo（camelCase）→ 后端 ApiKeyInput wire 形态（snake_case） */
function toApiKeyInputWire(key: ApiKeyInfo): Record<string, unknown> {
  const wire: Record<string, unknown> = {}
  if (key.key != null) wire.key = key.key
  if (key.keyId != null) wire.key_id = key.keyId
  if (key.enabled !== undefined) wire.enabled = key.enabled
  if (key.priority !== undefined) wire.priority = key.priority
  return wire
}

/** 前端 camelCase 创建参数 → 后端 snake_case 请求体 */
function buildCreateBody(payload: ModelProviderCreate): Record<string, unknown> {
  const body: Record<string, unknown> = { name: payload.name, type: payload.type }
  if (payload.baseUrl != null) body.base_url = payload.baseUrl
  if (payload.apiKeys) body.api_keys = payload.apiKeys.map(toApiKeyInputWire)
  if (payload.autoFetch !== undefined) body.auto_fetch = payload.autoFetch
  return body
}

/** 前端 camelCase 更新参数 → 后端 snake_case 请求体（仅显式传入的字段） */
function buildUpdateBody(payload: ModelProviderUpdate): Record<string, unknown> {
  const body: Record<string, unknown> = {}
  if (payload.name !== undefined) body.name = payload.name
  if (payload.type !== undefined) body.type = payload.type
  if (payload.baseUrl !== undefined) body.base_url = payload.baseUrl
  if (payload.apiKeys !== undefined) body.api_keys = payload.apiKeys.map(toApiKeyInputWire)
  if (payload.models !== undefined) {
    body.models = payload.models.map((m) => ({
      model_id: m.modelId,
      enabled: m.enabled,
      supports_1m_context: m.supports1mContext,
    }))
  }
  return body
}

/** GET /model-providers —— 提供商列表摘要（不含任何 Key 信息） */
export async function listModelProviders(): Promise<ProviderSummary[]> {
  const { data } = await client.get<Record<string, unknown>[]>('/model-providers')
  return data.map(toProviderSummary)
}

/** POST /model-providers —— 创建提供商（默认自动获取模型列表，失败不阻断创建） */
export async function createModelProvider(
  payload: ModelProviderCreate,
): Promise<ModelProviderCreateResponse> {
  const { data } = await client.post<Record<string, unknown>>('/model-providers', buildCreateBody(payload))
  return {
    provider: toModelProvider(data.provider as Record<string, unknown>),
    ...(data.auto_fetch != null
      ? { autoFetch: toFetchResult(data.auto_fetch as Record<string, unknown>) }
      : {}),
    ...(data.message != null ? { message: data.message as string } : {}),
  }
}

/** GET /model-providers/{provider_id} —— 提供商详情（api_keys 脱敏） */
export async function getModelProvider(providerId: string): Promise<ModelProvider> {
  const { data } = await client.get<Record<string, unknown>>(`/model-providers/${providerId}`)
  return toModelProvider(data)
}

/** PATCH /model-providers/{provider_id} —— 更新提供商（名称/base_url/api_keys/models） */
export async function updateModelProvider(
  providerId: string,
  payload: ModelProviderUpdate,
): Promise<ModelProvider> {
  const { data } = await client.patch<Record<string, unknown>>(
    `/model-providers/${providerId}`,
    buildUpdateBody(payload),
  )
  return toModelProvider(data)
}

/** DELETE /model-providers/{provider_id} —— 删除提供商 */
export async function deleteModelProvider(providerId: string): Promise<void> {
  await client.delete(`/model-providers/${providerId}`)
}

/** POST /model-providers/{provider_id}/fetch-models —— 获取模型列表（合并去重所有启用 Key） */
export async function fetchModels(providerId: string): Promise<ModelFetchResult> {
  const { data } = await client.post<Record<string, unknown>>(
    `/model-providers/${providerId}/fetch-models`,
  )
  return toFetchResult(data)
}

/** POST /model-providers/{provider_id}/detect —— 检测所有 Key 连接状态 */
export async function detectProvider(providerId: string): Promise<KeyDetectResult[]> {
  const { data } = await client.post<Record<string, unknown>[]>(`/model-providers/${providerId}/detect`)
  return data.map(toDetectResult)
}

/** POST /model-providers/{provider_id}/keys/{key_id}/detect —— 检测单个 Key 连接状态 */
export async function detectKey(providerId: string, keyId: string): Promise<KeyDetectResult> {
  const { data } = await client.post<Record<string, unknown>>(
    `/model-providers/${providerId}/keys/${keyId}/detect`,
  )
  return toDetectResult(data)
}

/** 与代码库其他 API 模块一致的对象命名空间（自由函数 + 对象双出口） */
export const modelProvidersApi = {
  listModelProviders,
  createModelProvider,
  getModelProvider,
  updateModelProvider,
  deleteModelProvider,
  fetchModels,
  detectProvider,
  detectKey,
}
